### PR TITLE
Refactor index.go

### LINKED ISSUE #191: Refactor index.go data operations into dao
**Is your feature request related to a problem? Please describe.**
Loading the courses in [index.go](https://github.com/joschahenningsen/TUM-Live/blob/dev/web/index.go) could be faster.

**Describe the solution you'd like**
Prepare the data in the dao accordingly to avoid sorting and filtering in `index.go`.

### PR DESCRIPTION
**Pull request by MatthiasReumann**
- Split up function into smaller ones 
- Added function to DAO for public and loggedIn courses with caching
- For 'LoadLivestreams' I added the function 'GetCurrentLiveNonHidden' which checks if the course associated with the stream isn't hidden. (Joins in DAO functions?)

### CODE REVIEW

File: model/course.go [hunk 1]
```diff
@@ -158,4 +158,9 @@ func (c Course) GetRecordings() []Stream {
...
 	}
 	return recordings
-}
\ No newline at end of file
+}
+
+// IsHidden returns true if visibility is set to 'hidden' and false if not
+func (c Course) IsHidden() bool {
+	return c.Visibility == "hidden"
+}
```
Comment (joschahenningsen) [inline model/course.go:176]: So much better than checking this in the .gohtml :100:

File: web/index.go [hunk 2]
```diff
@@ -29,62 +100,51 @@ func MainPage(c *gin.Context) {
...
-	var tumLiveContext tools.TUMLiveContext
-	tumLiveContextQueried, found := c.Get("TUMLiveContext")
-	if found {
-		tumLiveContext = tumLiveContextQueried.(tools.TUMLiveContext)
-		indexData.TUMLiveContext = tumLiveContext
-	} else { //TODO log
-		c.AbortWithStatus(http.StatusInternalServerError)
-		return
-	}
+func (d *IndexData) SetYearAndTerm(c *gin.Context) {
```
Comment (joschahenningsen) [inline web/index.go:117]: Can you add some comments to the new exported functions?

File: dao/courses.go [hunk 3]
```diff
@@ -137,6 +157,14 @@ func GetCourseById(ctx context.Context, id uint) (course model.Course, err error
 	return foundCourse, dbErr
 }
 
+func GetNonHiddenCourseById(ctx context.Context, id uint) (course model.Course, err error) {
+	var foundCourse model.Course
+	dbErr := DB.Preload("Streams.Stats").Preload("Streams.Files").Preload("Streams", func(db *gorm.DB) *gorm.DB {
```
Comment (joschahenningsen) [inline dao/courses.go]: I think this doesn't need to preload the stats, as the function is not used in the admin tab.

Comment (joschahenningsen) [reply dao/courses.go]: BTW, where is it used?

File: dao/courses.go [hunk 4]
```diff
@@ -96,15 +96,35 @@ func GetPublicCourses(year int, term string) (courses []model.Course, err error)
...
+		return cachedCourses.([]model.Course), err
+	}
+	var publicCourses []model.Course
+
+	err = DB.Preload("Streams", func(db *gorm.DB) *gorm.DB {
+		return db.Order("start asc")
+	}).Find(&publicCourses,
+		"(visibility = 'public' OR visibility = 'loggedin') AND teaching_term = ? AND year = ?", term, year).Error
+	if err == nil {
+		Cache.SetWithTTL(fmt.Sprintf("publicAndLoggedInCourses%v%v", year, term), publicCourses, 1, time.Minute)
```
Comment (joschahenningsen) [inline dao/courses.go]: nit: I did this wrong as well and this works just fine, but the right format string would be `publicAndLoggedInCourses%d%s`

**Review by joschahenningsen — CHANGES_REQUESTED** [review]
This is much better!

File: dao/courses.go (hunk 3 above)
Comment (MatthiasReumann) [reply dao/courses.go]: Nowhere actually.😅 Interesting that the test didn't fail for the unused function.

Comment (joschahenningsen) [reply dao/courses.go]: It's exported (capitalized), so the linter assumes we might want other projects to import it :smile:

**Review by joschahenningsen — APPROVED** [review]
Thanks :)

**Pull request merged**